Enhance SelectionSet annotation to ignore fields in the return type #27 (#39)
paypal/mocca: docs/END_USER_DOCUMENT.md

#### 4.5.2 Using Mocca SelectionSet annotation

There might be cases when the application has a DTO it could use as return type for a GraphQL operation, however, not all its fields are really required to be returned from the server. In this case, the annotation `com.paypal.mocca.client.annotation.SelectionSet` can be used to customize the selection set, letting the application specify exactly what the data it would like to receive or ignore.

> Mocca SelectionSet supports two options:

Reviewer: ```suggestion
Mocca `SelectionSet` supports two options:
```

- value
- ignore

> SelectionSet behaviour:

Reviewer: ```suggestion
`SelectionSet` behaviour:
```

> - If annotation is present and its value attribute is set, Mocca automatic selection set resolution is turned off, and SelectionSet value is used to define the selection set. In this case if ignore value is also set, then that is not used by Mocca, and a warning is logged.

Reviewer: ```suggestion
- If annotation is present and its `value` attribute is set, Mocca automatic selection set resolution is turned off, and `SelectionSet` `value` is used to define the selection set. In this case if `ignore` is also set, then that is not used by Mocca, and a warning is logged.
```

> - If annotation is present and its `value` attribute is set, Mocca automatic selection set resolution is turned off, and `SelectionSet` `value` is used to define the selection set. In this case if `ignore` value is also set, then that is not used by Mocca, and a warning is logged.

Reviewer: ```suggestion
- If annotation is present and its `value` attribute is set, Mocca automatic selection set resolution is turned OFF, and `SelectionSet` `value` is used to define the selection set. In this case if `ignore` value is also set, then that is not used by Mocca, and a warning is logged.
```

> - If annotation is present, its value attribute is NOT set, but ignore is, then Mocca automatic selection set resolution is turned ON, and SelectionSet ignore is used to pick which response DTO fields to ignore from the selection set.

Reviewer: ```suggestion
- If annotation is present, its value attribute is NOT set, but ignore is, then Mocca automatic selection set resolution is turned ON, and `SelectionSet` `ignore` is used to pick which response DTO fields to ignore from the selection set.
```

> - If annotation is present, its value attribute is NOT set, but ignore is, then Mocca automatic selection set resolution is turned ON, and `SelectionSet` `ignore` is used to pick which response DTO fields to ignore from the selection set.

Reviewer: ```suggestion
- If annotation is present, its value attribute is NOT set, but `ignore` is, then Mocca automatic selection set resolution is turned ON, and `SelectionSet` `ignore` is used to pick which response DTO fields to ignore from the selection set.
```

> - If annotation is present and both value and ignore attributes are NOT set, then a MoccaException is thrown.

Reviewer: ```suggestion
- If annotation is present and both value and ignore attributes are NOT set, then a `MoccaException` is thrown.
```

> - If annotation is present and both value and ignore attributes are NOT set, then a `MoccaException` is thrown.

Reviewer: ```suggestion
- If annotation is present and both `value` and `ignore` attributes are NOT set, then a `MoccaException` is thrown.
```

``` java
@Query

> @SelectionSet(value="{id, title}")

Reviewer: The `value` attribute in annotations is "especial". When it is the only parameter set in an annotation, it doesn't have to be mentioned. So, this line was actually right, and was done that way intentionally, since it is less verbose. Having said that, please undo this change. Thanks.

}
```

> If application wants to use the `Book` DTO as return type, but it wants to ignore the title field only. Notice `SelectionSet` ignore must be sent as an array of String values.

Reviewer: ```suggestion
The example below shows an application using `Book` DTO as return type, but ignoring only the `title` field. Notice multiple values can be set to `SelectionSet` `ignore` as an array of `String` values.
```

##### GraphQL operation definition in the client API

``` java
@Query

> @SelectionSet(ignore={"title"})

Reviewer: ```suggestion
@SelectionSet(ignore="title")
```

Reviewer: Array fields in annotations (of any type, including Strings) can be set as one item, without curly braces, when the intention is to set only one value.
